superset/db_engine_specs/trino.py

```diff
@@ -149,14 +151,26 @@ def get_tracking_url(cls, cursor: Cursor) -> str | None:
 
     @classmethod
     def handle_cursor(cls, cursor: Cursor, query: Query, session: Session) -> None:
-        if tracking_url := cls.get_tracking_url(cursor):
-            query.tracking_url = tracking_url
+        """
+        Handle a trino client cursor.
+
+        WARNING: if you execute a query, it will block until complete and you
+        will not be able to handle the cursor until complete. Use
+        `execute_with_cursor` instead, to handle this asynchronously.
+        """
 
         # Adds the executed query id to the extra payload so the query can be cancelled
-        query.set_extra_json_key(
-            key=QUERY_CANCEL_KEY,
-            value=(cancel_query_id := cursor.stats["queryId"]),
-        )
+        # Since we're racing with the query execution, we need to poll for it
+        # TODO: do we need a max wait time here? Or rely on conn timeout?
+        while "queryId" not in cursor.stats:
```

Comment:
After internal peer review, I'm moving this loop out to the `execute_with_cursor` method and only enter here once we see either the query finish or the query ID appear; that's because it's possible the connection will fail, query ID will never appear, and this loop will just continue until the connection times out.

Reply:
Similarly we can just use `cursor.query_id` which is a bit cleaner.